src/remote_tables/pushdown_visitor.rs

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+}
+
+pub trait FilterPushdownVisitor {
+    fn scalar_value_to_sql(&self, value: &ScalarValue) -> Option<String> {
+        match value {
+            ScalarValue::Utf8(Some(val)) => Some(format!("'{}'", val)),
+            _ => Some(format!("{}", value)),
+        }
+    }
+
+    fn op_to_sql(&self, op: Operator) -> Option<String> {
+        Some(op.to_string())
+    }
+}
+
+impl<T: FilterPushdownVisitor> ExpressionVisitor for FilterPushdown<T> {
+    fn pre_visit(mut self, expr: &Expr) -> Result<Recursion<Self>> {
+        match expr {
+            Expr::Column(_) | Expr::Literal(_) => {}
+            Expr::BinaryExpr(BinaryExpr { op, .. }) => {
+                // Check if operator pushdown supported; left and right expressions will be checked
+                // through further recursion.
+                if self.source.op_to_sql(*op).is_none() {
+                    return Ok(Recursion::Stop(self));
+                }
+            }
+            _ => {
+                // Expression is not supported, no need to visit any remaining nodes
+                self.pushdown_supported = false;
+                return Ok(Recursion::Stop(self));
+            }
+        };
+        Ok(Recursion::Continue(self))
+    }
+
+    fn post_visit(mut self, expr: &Expr) -> Result<Self> {
+        match expr {
+            Expr::Column(col) => self.sql_exprs.push(col.name.clone()),
+            Expr::Literal(val) => {
+                let sql_val = self.source.scalar_value_to_sql(val).ok_or_else(|| {
+                    DataFusionError::Execution(format!(
+                        "Couldn't convert ScalarValue {:?} to a compatible one for the remote system",
+                        val,
+                    ))
+                })?;
+                self.sql_exprs.push(sql_val)
+            }
+            Expr::BinaryExpr(be @ BinaryExpr { .. }) => {
+                // The visitor has been through left and right sides in that order, so the topmost
+                // item on the SQL expression stack is the right expression
+                let mut right_sql = self.sql_exprs.pop().unwrap_or_else(|| {
+                    panic!("Missing right sub-expression of {}", expr)
+                });
+                let mut left_sql = self
+                    .sql_exprs
+                    .pop()
+                    .unwrap_or_else(|| panic!("Missing left sub-expression of {}", expr));
+
+                // Similar as in Display impl for BinaryExpr: since the Expr has an implicit operator
+                // precedence we need to convert it to an explicit one using extra parenthesis if the
+                // left/right expression is also a BinaryExpr of lower operator precedence.
+                if let Expr::BinaryExpr(right_be @ BinaryExpr { .. }) = &*be.right {
+                    let p = right_be.precedence();
+                    if p == 0 || p < be.precedence() {
+                        right_sql = format!("({})", right_sql)
```

Comment:
You could also always use brackets for order of operations and not bother with precedence (and potential bugs), since nobody will probably be looking at the actual generated SQL (we're also assuming here that DataFusion's operator precedence rules are the same as the remote database).

Reply:
Makes sense, though note that this still might introduce bugs due to operator precedence mismatch (but low risk)